src/openms/source/ANALYSIS/ID/AhoCorasickDA.cpp

```diff
@@ -0,0 +1,820 @@
+    {
+      UInt32 idx = static_cast<UInt32>(std::distance(sorted_idx_.begin(), std::find(sorted_idx_.begin(), sorted_idx_.end(), peptide_index)));
+      if (idx+1 < sequences_.size() && sequences_[sorted_idx_[idx]] == sequences_[sorted_idx_[idx+1]])
+      {
+        peptide_index = sorted_idx_[idx+1];
+        pos_in_protein = pos_in_protein;
+        return true;
+      }
+    }
+
+    while (*protein_ != '\0')
+    {
+      if (retrieval_())
+      {
+        peptide_index = output_[node_pos_];
+        pos_in_protein = prot_pos_ + 1 - sequences_[peptide_index].size();
+        protein_++;
+        prot_pos_++;
+        Int32 next_node = getNode_(node_pos_, code_(*protein_));
+        if (next_node == 0)
+        {
+          node_pos_ = getSupplyLink_(node_pos_);
+          // TODO for the retrieval it must be the next node in Tail
+          if (node_pos_ < 0 )
+          {
+            node_pos_ = node_pos_-1;
+          }
+        }
+        else
+        {
+          node_pos_ = next_node;
+        }
+        return true;
+      }
+      else
+      {
+        node_pos_ = getSupplyLink_(node_pos_);
+        // TODO for the retrieval it must be the next node in Tail
+        if (node_pos_ < 0 )
+        {
+          node_pos_ = node_pos_-1;
+        }
+      }
+    }
+    return false;
+  }
+
+
+  void AhoCorasickDA::construct_()
+  {
+    if (sequences_.empty())
+    {
+      throw Exception::InvalidParameter(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No sequences given");
+    }
+
+    // sort strings
+
+    // index vector of sorted pattern
+    sorted_idx_.resize(sequences_.size());
+    std::iota(sorted_idx_.begin(), sorted_idx_.end(), 0);
+    std::sort(sorted_idx_.begin(), sorted_idx_.end(), [this](UInt32 idx1, UInt32 idx2){return sequences_[idx1].compare(sequences_[idx2]) < 0;});
+
+    // unique vector
+    std::unique_copy(sorted_idx_.begin(), sorted_idx_.end(), std::back_inserter(unique_idx_), [this](UInt32 idx1, UInt32 idx2){return sequences_[idx1]==sequences_[idx2];});
+
+    //DA
+
+    Size init_capa = 1;
+    while (init_capa < unique_idx_.size())
+    {
+      init_capa <<= 1;
+    }
+
+    bc_.reserve(init_capa);
+    build_info_.reserve(init_capa);
+    edges_.reserve(22); // 23 with node method
+
+
+    expand_();
+    arrangeEmpty_(0);
+    buildTrie_(0, unique_idx_.size(), 0, 0);
+    constructFailure_();
+  }
+
+  void AhoCorasickDA::expand_()
+  {
+    auto old_size = static_cast<uint32_t>(bc_.size());
+    auto new_size = old_size + 256;
+
+
+    // Exception size of DA becomes to large to store
+    if (new_size >= 1<<22)
+    {
+      throw Exception::InvalidSize(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, new_size);
+    }
+
+
+    // add new empty nodes
+    for (uint32_t i = 0; i < 256; ++i)
+    {
+      bc_.push_back(BC{});
+      build_info_.push_back(BuildInformation{});
+    }
+
+
+    // set linkage
+    for (auto i = old_size; i < new_size; i++)
+    {
+      build_info_[i].prev = i - 1;
```

Comment:
if you are iterating over the elements you created above, then why not use a CTor like
`build_info_.emplace_back(i-1, i+1)`

Reply:
less good 
`build_info_.push_back(BuildInformation(i-1, i+1));`